### PR TITLE
Modify S3257: Migrate to LayC

### PR DESCRIPTION
**Pull request by victor-diez-sonarsource**
## Review

A dedicated reviewer checked the rule description successfully for:

- [ ] logical errors and incorrect information
- [ ] information gaps and missing content
- [ ] text style and tone
- [ ] PR summary and labels follow [the guidelines](https://github.com/SonarSource/rspec/#to-modify-an-existing-rule)

### CODE REVIEW

File: rules/S3257/javascript/rule.adoc [hunk 1]
```diff
@@ -1,10 +1,18 @@
 == Why is this an issue?
 
-Unnecessarily verbose declarations and initializations make it harder to read the code, and should be simplified. Specifically, primitive (``++number++``, ``++string++``, ``++boolean++`` and others) types should be omitted from variable and parameter declaration when they can be easily inferred from the initialized or defaulted value.
+TypeScript has a feature called "type inference" which allows it to automatically infer the type of a variable based on its initial value. This means that if you initialize a variable with a certain type of value, TypeScript will assume that this variable should always hold that type of value.
```
Comment (yassin-kammoun-sonarsource) [inline rules/S3257/javascript/rule.adoc:3]:
```suggestion
TypeScript supports type inference, a mechanism that automatically infers the type of a variable based on its initial value. This means that if you initialize a variable with a particular value, TypeScript will assume that this variable should always hold that type of value.
```

File: rules/S3257/javascript/rule.adoc [hunk 2]
```diff
@@ -1,10 +1,18 @@
...
 
-=== Noncompliant code example
+Unnecessarily verbose declarations and initializations make it harder to read the code and should be simplified. Specifically, primitive (``++number++``, ``++string++``, ``++boolean++`` and others) types should be omitted from variable and parameter declaration when they can be easily inferred from the initialized or defaulted value.
 
-[source,javascript]
+== How to fix it
+
+When the type can be easily inferred from the context, omit the type declaration to make the code more concise and readable.
```
Comment (yassin-kammoun-sonarsource) [inline rules/S3257/javascript/rule.adoc:9]:
```suggestion
Omit explicit type annotations in declarations whenever the type can be inferred from the context.
```

File: rules/S3257/javascript/rule.adoc [hunk 3]
```diff
@@ -1,10 +1,18 @@
 == Why is this an issue?
 
-Unnecessarily verbose declarations and initializations make it harder to read the code, and should be simplified. Specifically, primitive (``++number++``, ``++string++``, ``++boolean++`` and others) types should be omitted from variable and parameter declaration when they can be easily inferred from the initialized or defaulted value.
+TypeScript has a feature called "type inference" which allows it to automatically infer the type of a variable based on its initial value. This means that if you initialize a variable with a certain type of value, TypeScript will assume that this variable should always hold that type of value.
 
-=== Noncompliant code example
+Unnecessarily verbose declarations and initializations make it harder to read the code and should be simplified. Specifically, primitive (``++number++``, ``++string++``, ``++boolean++`` and others) types should be omitted from variable and parameter declaration when they can be easily inferred from the initialized or defaulted value.
```
Comment (yassin-kammoun-sonarsource) [inline rules/S3257/javascript/rule.adoc:5]:
```suggestion
Unnecessarily verbose declarations and initializations make it harder to read the code and should be simplified. Specifically, primitive (``++number++``, ``++string++``, ``++boolean++`` and others) types should be omitted from variable and parameter declarations when they can be easily inferred from the initialized or defaulted value.
```

Comment (yassin-kammoun-sonarsource) [inline rules/S3257/javascript/rule.adoc:5]: If the rule isn't limited to primitive types, I would reword this as follows:

```suggestion
Unnecessarily verbose declarations and initializations make it harder to read the code and should be simplified. Therefore, type annotations should be omitted from variable and parameter declarations when they can be easily inferred from the initialized or defaulted value.
```

**Review by yassin-kammoun-sonarsource — APPROVED** [review]

File: rules/S3257/csharp/rule.adoc [hunk 4]
```diff
@@ -1,9 +1,10 @@
 == Why is this an issue?
 
-Unnecessarily verbose declarations and initializations make it harder to read the code, and should be simplified.
+In C#, the type of a variable can be inferred by the compiler at compile time. This is done using the `var` keyword. The use of `var` allows you to not repeat a type name in a variable declaration and object instantiation.
```
Comment (zsolt-kolbay-sonarsource) [inline rules/S3257/csharp/rule.adoc:3]:
```suggestion
In C#, the type of a variable can often be inferred by the compiler. The use of the [var keyword](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/implicitly-typed-local-variables) allows you to avoid repeating the type name in a variable declaration and object instantiation because the declared type can often be inferred by the compiler.
```

I changed this for two reasons:
1. Type inference doesn't only happen when you use the var keyword, e.g. also when passing parameters to methods.
2. The compiler can't always infer the type (hence I added the word _often_), e.g. you'd get an error for this declaration: `var x = null;`

**Review by zsolt-kolbay-sonarsource — APPROVED** [review]
LGTM
